pkg/gardenlet/controller/shoot/seed_registration_control.go

```diff
@@ -113,884 +68,298 @@ func (c *Controller) seedRegistrationUpdate(oldObj, newObj interface{}) {
 	}
 
-	if shoot.DeletionTimestamp == nil && shootedSeedConfig != nil {
-		if shoot.Generation != shoot.Status.ObservedGeneration || shoot.Status.LastOperation == nil || shoot.Status.LastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
-			shootLogger.Infof("[SHOOTED SEED REGISTRATION] Waiting for shoot %s to be reconciled before registering it as seed", shoot.Name)
-			return reconcile.Result{
-				RequeueAfter: 10 * time.Second,
-			}, nil
-		}
-
-		seedClient, err := c.clientMap.GetClient(ctx, keys.ForSeedWithName(*shootObj.Spec.SeedName))
-		if err != nil {
-			return reconcile.Result{}, fmt.Errorf("failed to get seed client: %w", err)
-		}
-
-		if shoot.Spec.DNS == nil || shoot.Spec.DNS.Domain == nil {
-			return reconcile.Result{}, errors.New("cannot register Shoot as Seed if it does not specify a domain")
-		}
-
-		if shootedSeedConfig.NoGardenlet {
-			shootLogger.Infof("[SHOOTED SEED REGISTRATION] Registering %s as seed as configuration says that no gardenlet is desired", shoot.Name)
-			if err := registerAsSeed(ctx, gardenClient.Client(), seedClient.Client(), shoot, shootedSeedConfig); err != nil {
-				message := fmt.Sprintf("Could not register shoot %q as seed: %+v", shoot.Name, err)
-				shootLogger.Errorf(message)
-				c.recorder.Event(shoot, corev1.EventTypeWarning, "SeedRegistration", message)
-				return reconcile.Result{}, err
-			}
-		} else {
-			shootedSeedClient, err := c.clientMap.GetClient(ctx, keys.ForShoot(shoot))
-			if err != nil {
-				return reconcile.Result{}, fmt.Errorf("failed to get shooted seed client: %w", err)
-			}
-
-			shootLogger.Infof("[SHOOTED SEED REGISTRATION] Deploying gardenlet into %s which will register shoot as seed", shoot.Name)
-			if err := deployGardenlet(ctx, gardenClient, seedClient, shootedSeedClient, shoot, shootedSeedConfig, c.imageVector, c.config); err != nil {
-				message := fmt.Sprintf("Could not deploy Gardenlet into shoot %q: %+v", shoot.Name, err)
-				shootLogger.Errorf(message)
-				c.recorder.Event(shoot, corev1.EventTypeWarning, "GardenletDeployment", message)
-				return reconcile.Result{}, err
-			}
-		}
-	} else {
-		shootLogger.Infof("[SHOOTED SEED REGISTRATION] Deleting `Seed` object for %s", shoot.Name)
-		if err := deregisterAsSeed(ctx, gardenClient, shoot); err != nil {
-			message := fmt.Sprintf("Could not deregister shoot %q as seed: %+v", shoot.Name, err)
-			shootLogger.Errorf(message)
-			c.recorder.Event(shoot, corev1.EventTypeWarning, "SeedDeletion", message)
-			return reconcile.Result{}, err
-		}
+	exists, isOwnedBy, err := isManagedSeedOwnedBy(ctx, gardenClient.Client(), shoot)
+	if err != nil {
+		message := fmt.Sprintf("Could not get ManagedSeed object for shoot %q: %+v", kutil.ObjectName(shoot), err)
+		shootLogger.Errorf(message)
+		c.recorder.Event(shoot, corev1.EventTypeWarning, "ManagedSeedGet", message)
+		return reconcile.Result{}, err
+	}
+	if exists && !isOwnedBy {
```

Comment:
What about exiting early on `!exists && shootedSeed == nil`?
Currently, on startup of the gardenlet, I get a log message like this for all shoots in the `garden` namepsace, that don't even have the `use-as-seed` annotation, which is quite confusing:
```text
msg="[SHOOTED SEED REGISTRATION] Deleting ManagedSeed object for shoot garden/seed-gcp" shoot=garden/seed-gcp
```

Reply:
Absolutely, I also noticed this, will add this to the above if.

Reply:
Did it slightly differently, but should work as expected.